### DISCUSSION TITLE
Rough WASM/WebGPU ETA?

### DISCUSSION THREAD

**Discussion by josephrocca** [iree-org/iree#7757] [Q&A] [answered]
Just wondering whether there is any rough ETA on WASM and WGSL code generation? I'm guessing that WASM will come first since WebGPU hasn't even been standardized yet? Even a super rough ETA would be appreciated (e.g. "probably 2022/23 for wasm") if that's all that can be provided at this stage. Thanks!

**Comment by stellaraccident** [comment, marked as answer]
Here's the status as I know it:

* We're in close contact with the WebGPU team at Google (which includes our delegates working on the standard) and ask the question each quarter of "is it time yet" -- basically aiming for it being at a point of stability in Chrome so that we can build something concrete without getting too into the weeds. We expect that a systematic translation from SPIR-V to WGSL will be possible.
* There are two use cases for WASM:
  * As a general purpose (non-browser based) sandbox target: We have an internal prototype of some extensions to WASM runtimes to enable the necessary host memory interface that would yield the best implementation on our side. This just got brought up a few weeks ago and I don't have an ETA yet on contributing the necessary upstream changes, but didn't expect it to be extremely burdensome.
  * For use in a web browser: We're not actively working on this right now but have had various demos of it in the past. Roughly looking to do it right in 2022.

Since we are still getting the core platforms in good shape, and since we have evidence that WebAssembly/WebGPU can build on those investments, we've been putting both into the category of "do it right/do it once when the necessary pre-requisites are met to make it great." It looks like both will line up for 2022, but I don't have a firm ETA.

**Reply by josephrocca** [reply]
Thank you! For others hitting this discussion thread wondering specifically about WebGPU rollout (in Chromium at least), [this comment](https://github.com/gpuweb/gpuweb/discussions/2003#discussioncomment-1091593) by Kangz provides some timing estimates.
